node/node_ping_controller.go

```diff
@@ -0,0 +1,104 @@
+package node
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"github.com/virtual-kubelet/virtual-kubelet/trace"
+	"golang.org/x/sync/singleflight"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type nodePingController struct {
+	nodeProvider       NodeProvider
+	pingInterval       time.Duration
+	firstPingCompleted chan struct{}
+	pingTimeout        *time.Duration
+
+	// "Results"
+	sync.Mutex
+	result pingResult
+}
+
+type pingResult struct {
+	pingTime time.Time
+	error    error
+}
+
+func newNodePingController(node NodeProvider, pingInterval time.Duration, timeout *time.Duration) *nodePingController {
+	return &nodePingController{
+		nodeProvider:       node,
+		pingInterval:       pingInterval,
+		firstPingCompleted: make(chan struct{}),
+		pingTimeout:        timeout,
+	}
+}
+
+func (npc *nodePingController) run(ctx context.Context) {
+	const key = "key"
+	sf := &singleflight.Group{}
+
+	// 1. If the node is "stuck" and not responding to pings, we want to set the status
+	//    to that the node provider has timed out responding to pings
+	// 2. We want it so that the context is cancelled, and whatever the node might have
+	//    been stuck on uses context so it might be unstuck
+	// 3. We want to retry pinging the node, but we do not ever want more than one
+	//    ping in flight at a time.
+
+	mkContextFunc := context.WithCancel
+
+	if npc.pingTimeout != nil {
+		mkContextFunc = func(ctx2 context.Context) (context.Context, context.CancelFunc) {
+			return context.WithTimeout(ctx2, *npc.pingTimeout)
+		}
+	}
+
+	checkFunc := func(ctx context.Context) {
+		ctx, cancel := mkContextFunc(ctx)
+		defer cancel()
+		ctx, span := trace.StartSpan(ctx, "node.pingLoop")
+		defer span.End()
+		doChan := sf.DoChan(key, func() (interface{}, error) {
```

Comment:
So the problem here is we still can't cancel the goroutine.
Maybe let's add a doc to the `Ping` interface to say that implementations must cancel and return the context error when the context is cancelled.

I think I'd rather have bad implementations block here than leaking these goroutines.

Reply:
I totally agree with you abut leaking goroutines. In this implementation we will never have more than 1 ping call in-flight, the singeflight prohibits that.

All the cancel does is that it asks the Node implementation nicely to try to cancel the current singleflight.

Reply:
tl;dr: This will not leak more than one goroutine "worst case scenario"

Reply:
To think this through -- we cannot pre-empt the call to `Ping()`, _but_ we can make sure the previous call to ping returns before invoking another call to `Ping()`. The code isn't fully plumbed here, but my rough plan was:

1. We take a timestamp prior to pinging the node
2. When the ping is completed we record the timestamp when we started the ping
3. We give the user the option to base the lease renewal timestamp not on the node ping response timestamp.